File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int maxArea = 0;
        int i = 0, j=heights.length-1;

        while(i<j) {
            int area = Math.min(heights[i],heights[j]) * (j-i);
            maxArea = Math.max(maxArea, area);
            if(heights[i] < heights[j]) {
                i++;
            } else {
                j--;
            }
        }
        return maxArea;
    }
}

File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-0.js
class Solution {
    /**
     * @param {number[]} weights
     * @param {number} days
     * @return {number}
     */
    shipWithinDays(weights, days) {
        let l = Math.max(...weights);
        let r = weights.reduce((a, b) => a + b, 0);
        let res = r;

        function canShipWithCapacity(cap) {
            let ships = 1;
            let currCap = cap;

            for(let i = 0; i < weights.length; i++) {
                if(currCap - weights[i] < 0) {
                    ships++;
                    currCap = cap;
                }
                currCap -= weights[i];
            }

            return ships <= days;
        }

        while(l <= r) {
            let cap = Math.floor((l + r) / 2);

            if (canShipWithCapacity(cap)) {
                r = cap - 1;
                res = Math.min(res, cap);
            } else {
                l = cap + 1;
            }
        }

        return res;
    }
}
